vllm/engine/async_llm_engine.py

```diff
@@ -296,6 +360,66 @@ async def step_async(
 
         return request_outputs
 
+    def _has_remaining_steps(
+        self, seq_group_metadata_list: Optional[List[SequenceGroupMetadata]]
+    ) -> bool:
+        if not self.scheduler_config.is_multi_step:
+            return False
+
+        if seq_group_metadata_list is None:
+            return False
+        # TODO(will) this is a sanity check for nowto make sure that all the
+        # seqs are on the same steps. Eventually we will want to do some sort of
+        # dynamic scheduling when doing multi-step decoding.
+        if len(seq_group_metadata_list) == 0:
+            return False
+        steps_remaining = [
+            seq_group.state.remaining_steps
+            for seq_group in seq_group_metadata_list
+        ]
+        if steps_remaining.count(steps_remaining[0]) != len(steps_remaining):
+            raise AssertionError(("All running sequence groups should "
+                                  "have the same remaining steps."))
```

Comment:
Might be more efficient.
```suggestion
        ref_remaining_steps = seq_group_metadata_list[0].state.remaining_steps
        if any([seq_group.state.remaining_steps != ref_remaining_steps for seq_group in seq_group_metadata_list[1:]]):
            raise AssertionError(("All running sequence groups should "
                                  "have the same remaining steps."))
```

Reply:
[22daa7e](https://github.com/vllm-project/vllm/pull/7000/commits/22daa7ed840a03f12ff89d933c7035176bd8ba1d)